### ISSUE TITLE
wufuc is incompatible with UpdatePack7R2

### ISSUE THREAD

**Issue by numpa** [zeffy/wufuc#100]
Getting Unsupported Hardware message on Core i5-760 CPU with Windows 7 Home Premium. Tried to install wufuc v0.7.1.81 both ways, msi and bat with admin rights, no changes. UAC is turned on. Thanks in advance.

[logs.zip](https://github.com/zeffy/wufuc/files/1227767/logs.zip)

**Comment by zeffy** [comment]
That's really weird, according to the logs, your Windows Update service DLL is "wuaueng2.dll" rather than "wuaueng.dll" which is what it normally would be. Did you make this change, or is this unknown to you?

Would it be possible for you to get the SHA-1 hash and version number of that file? (It is located at `C:\Windows\System32\wuaueng2.dll`).

Thanks for reporting this!

**Comment by rautamiekka** [comment]
Google gives some dark results: a **crypto Trojan**. Certainly hope it's not that, but I've never seen that file myself either.

**Comment by zeffy** [comment]
Well, hopefully it's nothing like that... it's possible he just renamed the service DLL for some reason. But a VT scan should give a more clear picture of what we're looking at here.

**Comment by rautamiekka** [comment]
It's striking how this has _like never_ happened before, considering in Window$ land just anything happens.

**Comment by zeffy** [comment]
Indeed, this is a strange one. 

The logs also indicate that the byte pattern scan failed, which means it couldn't find the `IsDeviceServiceable` function. To the best of my knowledge, the pattern should work with all known versions of the wuaueng.dll file, so this is very peculiar.

**Comment by numpa** [comment]
Fortunately it's not "crypto Trojan". The system is freshly reinstalled and updated with newest version of UpdatePack7R2. System32 folder contains 2 files: wuaueng.dll and wuaueng2.dll. 

Info about wuaueng.dll:
File version: 7.6.7601.23806 (win7sp1_ldr.170510-0600)
Internal name: wuaueng.dll
Original filename: wuaueng.dll.mui
Language: Russian
Size: 2 651 136 bytes

Info about wuaueng2.dll:
File version: 7.6.7601.23453 (win7sp1_ldr.160513-1233)
Internal name: wuaueng.dll
Original filename: wuaueng.dll
Language: English (USA)
Size: 2 607 104 bytes

**Comment by zeffy** [comment]
wuaueng2.dll appears to be an older version that doesn't have the CPU check, thus wufuc can't patch it. 

Is it possible this "UpdatePack7R2" (which I've never heard of it before now, so please forgive my ignorance) is modifying your registry settings to point to this older wuaueng2.dll? If so, that would be the cause of the problem. Honestly, I would recommend just getting your updates directly from Microsoft to avoid issues like this.

First, you should run `sfc /SCANNOW` in an elevated command prompt and let it verify and fix any issues it finds with your Windows install.

If the SFC scan doesn't fix anything, I've exported my own wuauserv registry settings that point to wuaueng.dll, which you can import to hopefully fix your problem. **Only use this as a last resort, as I can't guarantee it will work correctly on your system. You should first export your own registry settings from `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\services\wuauserv\Parameters` before importing this.** [wuauserv_export.zip](https://github.com/zeffy/wufuc/files/1229549/wuauserv_export.zip)

**Comment by rautamiekka** [comment]
https://www.patreon.com/simplix -> https://translate.google.com/translate?sl=ru&tl=en&js=y&ie=UTF-8&u=http%3A%2F%2Fupdate7.simplix.info, it indeed incorporates some way to prevent the CPU check as mentioned near the bottom in the list of undesirable updates.

**Comment by zeffy** [comment]
Nice job spotting that! I didn't notice the first time I looked at the project page. I really wish people would just link to wufuc or include it in update packages like this instead of trying to come up with their own convoluted solution (which clearly doesn't even work) like hacking the registry and using older versions of wuaueng.dll 🙄

**Comment by numpa** [comment]
UpdatePack7R2 is just a pack of unmodified Windows and IE updates without telemetry/upgrade updates. And it installs updates with disabled telemetry and CPU checks. This pack is very useful when you install a fresh copy of Windows, because update it to the current state through Windows Update takes hours, sometimes days. This pack updates Windows to it's current state in about two hours with Core i5-760 with 4 GB RAM and 7200 RPM 7 years old HDD.

SFC scan hasn't found any errors.
I've checked the registry. The only difference with yours is that ServiceDll points to wuaueng2.dll instead of wuaueng.dll. But since it's Russian version of Windows, maybe it must be this way? I don't know why  wuaueng2.dll (which is English version of wuaueng.dll) is older than the Russian one.

**Comment by zeffy** [comment]
I believe UpdatePack7R2 is changing your registry settings to point to that older wuaueng2.dll in a naïve attempt to circumvent the CPU check. 

Under normal conditions, there is no wuaueng2.dll file, even on Russian versions of windows. Try changing the registry to correctly point to wuaueng.dll instead, and I think everything should get fixed.

**Comment by zeffy** [comment]
>If it helps, I can install this version of Windows in Hyper-V VM, check for wuaueng files, then update Windows through Windows Update and check wuaueng files again. Of course there will not be any Unsupported Hardware message and it may take a lot of time, because sometimes Windows may do "Checking for updates" from a couple of hours and up to two days.

Thanks for offering, but that's not necessary. I'm currently setting up a fresh Windows 7 VM and I'm going to install UpdatePack7R2 to see what changes it makes, and if I can make wufuc work in harmony with it, or at least how to fully revert the changes it makes to wuaueng.dll (if it does anything other than modify the registry and drop wuaueng2.dll).

**Comment by zeffy** [comment]
@numpa were you ever able to solve this by fixing your wuauserv registry entry?

**Comment by numpa** [comment]
Unfortunately I did not try it yet, I'm currently away and be back only tomorrow. So shall I try and report here, of course.
Were you any success with installing UpdatePack7R2 and fixing this issue it creates? Is it possible to fix it automatically by your program? I looked on how to send the bug report to UpdatePack developer and it seems the only way is through forum. I'll write there when I'm back, hope he'll see.

**Comment by zeffy** [comment]
@numpa Yes, I got UpdatePack7R2 installed in a VM and was able to investigate it further. On my VM, wuaueng2.dll got created after I ran it. But weirdly, it didn't install an older version of wuaueng.dll like it did for you, not really sure why.

Both my wuaueng.dll and wuaueng2.dll were the same version, except wuaueng2.dll was patched (using the same exact byte that I patch, no less... it's certainly possible he just happened to find the same method to bypass the cpu check that I did, but come on). Additionally, he changed a byte in the PE checksum header, for whatever reason. No idea why... he didn't even change it to be valid, lol.

That seems as far as he went to try to bypass it, so you should be able to remove his patch by fixing your registry to point to wuaueng.dll, and deleting his patched wuaueng2.dll.


Now, as far as automatically removing his patch when you install wufuc, as a user of both pieces of software, how do you think I should go about this? I could feasibly add a check in the installer that triggers a warning and gives you the option to remove his patch. Do you think that would be sufficient?

**Comment by numpa** [comment]
Finally got time to change the registry from wuaueng2.dll to wuaueng.dll, rebooted, and it worked! Everything is fine now, no Unsupported Hardware message and wufunc log tells "Patched Windows Update module!".
About fixing this issue by your patch. It seems to me that a warning about changes UpdatePack7R2 made to the system and changing registry back from wuaueng2.dll to wuaueng.dll should be enough. Of course it would be nice to offer fixing this automatically in the future in case user run UpdatePack7R2 again.

**Comment by zeffy** [comment]
@numpa That's great, I'm glad it's working for you now! 

After thinking about this for a while, I think I came up with a pretty clean solution. I can detect at run-time if `svchost.exe` is trying to load `wuaueng2.dll`, and just redirect it back to `wuaueng.dll` instead, if it exists. 

That way it should be fully compatible with UpdatePack7R2, and should the user choose to uninstall wufuc, the system would remain as it was prior.

**Comment by zeffy** [comment]
```
08/22/17 17:40:38 [896] helpers.c:94(suspend_other_threads): Suspended 31 other threads
08/22/17 17:40:38 [896] iathook.c:14(iat_hook): Modified import address: hmod=00000000FF3B0000, import=LoadLibraryExA, old addr=000000007756D860, new addr=000007FEFA5212C0
08/22/17 17:40:38 [896] iathook.c:14(iat_hook): Modified import address: hmod=00000000FF3B0000, import=LoadLibraryExW, old addr=0000000077575AE0, new addr=000007FEFA5214D0
08/22/17 17:40:38 [896] service.c:31(get_svcdllW): Service "wuauserv" DLL path: C:\Windows\system32\wuaueng2.dll
08/22/17 17:40:38 [896] helpers.c:102(resume_and_close_threads): Resumed 31 threads
08/22/17 17:40:41 [896] hooks.c:120(LoadLibraryExW_hook): UpdatePack7R2 compatibility fix: redirecting c:\windows\system32\wuaueng2.dll -> c:\windows\system32\wuaueng.dll
08/22/17 17:40:41 [896] hooks.c:126(LoadLibraryExW_hook): Loaded library: c:\windows\system32\wuaueng.dll
08/22/17 17:40:41 [896] hooks.c:161(PatchWUA): wuaueng!IsDeviceServiceable VA: 000007FEF65DF2E4
08/22/17 17:40:41 [896] hooks.c:174(PatchWUA): Patched boolean value #1: 000007FEF681E948 = 00000000
08/22/17 17:40:41 [896] hooks.c:183(PatchWUA): Successfully patched WUA module!
```

It works! Until I officially release v0.7.2 (I have a few more changes to make first), you can grab the latest builds from the appveyor artifacts to get these changes: https://ci.appveyor.com/project/zeffy/wufuc

**Comment by numpa** [comment]
Great to hear that, this is definitely much better solution. I'll wait for 0.7.2 release because I use update pack only for initial update, after it I use WU. Thank you very much for helping me with this problem!

**Comment by zeffy** [comment]
@numpa I released finally v0.8 yesterday, which included the fix for UpdatePack7R2. https://github.com/zeffy/wufuc/releases/latest
